olive/passes/onnx/gptq_utils.py

```diff
@@ -0,0 +1,243 @@
+
+        self.g_idx = g_idx.clone() if g_idx is not None else self.g_idx
+        intweight = self._quant_weight(layer_weight, scales, zeros, g_idx, need_transpose=False)
+
+        qzeros = zeros.t().contiguous()
+
+        if self.bits in [2, 4, 8]:
+            self._pack_for_even_bits(intweight, qzeros)
+        else:
+            self._pack_for_odd_bits(intweight, qzeros)
+
+    def _quant_weight(self, weight, scales, zeros, g_idx=None, need_transpose=True):
+        # quantize float weight to integer weight
+        scales = scales.t().contiguous()
+        zeros = zeros.t().contiguous()
+        g_idx = self.g_idx.long()
+        scale_zeros = zeros * scales
+        self.scales = scales.clone() if self.scales.sum() == 0 else self.scales
+
+        scale_mat = scales[g_idx]
+        scale_zeros_mat = scale_zeros[g_idx]
+        intweight_t = torch.round((weight.T + scale_zeros_mat) / scale_mat).to(torch.int)
+
+        if not need_transpose:
+            return intweight_t
+        return intweight_t.T
+
+    def _pack_for_even_bits(self, intweight, intzeros_T):
+        # pack multiple integer weights into one byte, this function is optimized for
+        # even bits
+        self.act_order = self.g_idx[: self.groupsize // self.bits].sum().item() != 0
+        assert self.act_order is False, "act_order=True is not supported"
+        intzeros_pt = intzeros_T.T.byte()
+        scales_pt = self.scales.T
+        intweight_pt = intweight.byte()
+        block_size = self.groupsize
+
+        rows, cols = intweight_pt.shape
+        blob_size = block_size // 2
+        k_blocks = (rows + block_size - 1) // block_size
+        padded_rows = k_blocks * block_size
+        pad_len = padded_rows - rows
+        if pad_len > 0:
+            intweight_pt = torch.nn.functional.pad(intweight_pt, (0, 0, 0, pad_len), "constant", 0)
+            intzeros_pt = torch.nn.functional.pad(intzeros_pt, (0, 0, 0, pad_len), "constant", 0)
+
+        intzeros_pt = (intzeros_pt[:, 0::2]) | (intzeros_pt[:, 1::2] << 4)
+        intzeros_pt = intzeros_pt.reshape(-1)
+
+        intweight_pt_t = intweight.T
+        intweight_pt_t = (intweight_pt_t[:, 0::2]) | (intweight_pt_t[:, 1::2] << 4)
+        intweight_pt_t = intweight_pt_t.reshape(cols, k_blocks, blob_size)
+
+        scales_pt = scales_pt.reshape(-1)
+
+        assert self.qweight.shape == intweight_pt_t.shape
+        assert self.qzeros.shape == intzeros_pt.shape
+
+        self.scales = scales_pt.contiguous()
+        self.qweight = intweight_pt_t.contiguous().byte()
+        self.qzeros = intzeros_pt.contiguous().byte()
+
+    def _pack_on_row_fast_248bit(self, pack_tensor, ori_int_tensor, bits):
+        if pack_tensor.shape[0] == ori_int_tensor.shape[0]:
+            ori_int_tensor = ori_int_tensor.T
+            pack_tensor = pack_tensor.T
+        compress_ratio = 32 // bits
+        i = 0
+        row = 0
+        while row < pack_tensor.shape[0]:
+            if bits in [2, 4, 8]:
```

Comment:
>if bits in [2, 4, 8]: [](http://example.com/codeflow?start=12&length=21)

validate the arguments before the loop?

if bits not in  [2,4,8]:
    raise ValueError("Only 2,4,8 bits are supported.")